include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -0,0 +1,835 @@
+                        }
+                        else
+                        {
+                            // multiple iterations
+                            // first 1 full
+                            auto __value = __tmp_ptr[__num_sub_groups_local * __sub_group_local_id + __offset];
+                            __sub_group_scan<__sub_group_size, /*__is_inclusive=*/true, /*__init_present=*/false>(
+                                __sub_group, __value, __reduce_op, __carry_last);
+
+                            // then some number of full iterations
+                            for (std::uint32_t __i = 1; __i < __pre_carry_iters - 1; __i++)
+                            {
+                                auto __reduction_idx = __i * __num_sub_groups_local * __sub_group_size +
+                                                       __num_sub_groups_local * __sub_group_local_id + __offset;
+                                __value = __tmp_ptr[__reduction_idx];
+                                __sub_group_scan<__sub_group_size, /*__is_inclusive=*/true, /*__init_present=*/true>(
+                                    __sub_group, __value, __reduce_op, __carry_last);
+                            }
+
+                            // final partial iteration
+                            auto __proposed_idx = (__pre_carry_iters - 1) * __num_sub_groups_local * __sub_group_size +
+                                                  __num_sub_groups_local * __sub_group_local_id + __offset;
+                            auto __remaining_elements =
+                                __elements_to_process - ((__pre_carry_iters - 1) * __sub_group_size);
+                            auto __reduction_idx = (__proposed_idx < __subgroups_before_my_group)
+                                                       ? __proposed_idx
+                                                       : __subgroups_before_my_group - 1;
+                            __value = __tmp_ptr[__reduction_idx];
+                            __sub_group_scan_partial<__sub_group_size, /*__is_inclusive=*/true,
+                                                     /*__init_present=*/true>(__sub_group, __value, __reduce_op,
+                                                                              __carry_last, __remaining_elements);
+                        }
+
+                        // steps 3+4) load global carry in from neighbor work-group
+                        //            and apply to local sub-group prefix carries
+                        auto __carry_offset = 0;
+
+                        std::uint8_t __iters =
+                            oneapi::dpl::__internal::__dpl_ceiling_div(__active_subgroups, __sub_group_size);
+
+                        std::uint8_t __i = 0;
+                        for (; __i < __iters - 1; ++__i)
+                        {
+                            __sub_group_partials[__carry_offset + __sub_group_local_id] = __reduce_op(
```

Comment:
We might want to create a helper for `__carry_offset + __sub_group_local_id` and reuse it in this loop and below.